File: backend/api/schemas/project.py
from datetime import datetime
from typing import Optional
from uuid import UUID
from pydantic import BaseModel, Field


class TechnologyBase(BaseModel):
    name: str
    slug: Optional[str] = None
    icon_url: Optional[str] = None


class ProjectBase(BaseModel):
    slug: str = Field(..., min_length=1, max_length=200)
    title: str = Field(..., min_length=1, max_length=200)
    description: str = Field(..., min_length=1)
    long_description: Optional[str] = None
    category: str = Field(..., min_length=1, max_length=100)
    status: str = Field(default="completed")
    featured: bool = Field(default=False)
    highlights: list[str] = Field(default_factory=list)
    technologies: list[str] = Field(default_factory=list)
    assets: list[dict] = Field(default_factory=list)
    github_url: Optional[str] = None
    demo_url: Optional[str] = None
    pdf_url: Optional[str] = None
    pdf_size: Optional[str] = None
    thumbnail_url: Optional[str] = None


class ProjectSchema(ProjectBase):
    id: Optional[UUID] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ProjectDetailSchema(ProjectSchema):
    pass


class ProjectCreateSchema(ProjectBase):
    pass

File: frontend/src/lib/schemas/project.ts
import { z } from 'astro/zod';

export const ProjectTechnologySchema = z.object({
  name: z.string().min(1),
});

export const ProjectAssetSchema = z.object({
  name: z.string().min(1),
  url: z.string().min(1),
});

export const ProjectManifestSchema = z.object({
  slug: z.string().min(1),
  title: z.string().min(1),
  description: z.string(),
  category: z.string().min(1),
  type: z.enum(['app', 'analysis']).default('app'),
  featured: z.boolean(),
  status: z.string(),
  technologies: z.array(ProjectTechnologySchema),
  githubUrl: z.string().url().nullable(),
  demoUrl: z.string().url().nullable(),
  notebookUrl: z.string().nullable().optional(),
  htmlUrl: z.string().nullable().optional(),
  assets: z.array(ProjectAssetSchema).optional(),
});

export const ProjectsManifestSchema = z.array(ProjectManifestSchema);
